Add row locking grammer.

chrisxu333 opened the pull request:
I added support for a full functional row locking grammar including:

- FOR (NO KEY) UPDATE (OF), FOR (KEY) SHARE (OF)
- NOWAIT and SKIP LOCKED specifier

Bouncner commented:
Wasn't aware of this feature. Just for reference: this is the postgres introducing this functionality https://github.com/postgres/postgres/commit/0ac5ad5134f2769ccbaefec73844f8504c4d6182

Bouncner commented:
@chrisxu333 Are you aware of any other database that uses a similar functionality and syntax? Or is this postgres-specific?

chrisxu333 commented:
@Bouncner Speaking of SELECT FOR UPDATE (OF) and SELECT FOR SHARE (OF), I believe most well-known databases support this, like MySQL, Oracle, CockroachDB, Postgres and many more, and I believe they all support NOWAIT and SKIP LOCKED flag. And for the NO KEY and KEY option, I know for sure that CockroachDB and Postgres support this but not sure about others.

mweisgut commented:
Related Postgres references:
- https://www.postgresql.org/docs/14/explicit-locking.html
- https://www.postgresql.org/docs/14/sql-select.html

mweisgut commented on src/parser/bison_parser.y:
```diff
@@ -1114,6 +1120,49 @@ alias : AS IDENTIFIER { $$ = new Alias($2); }
+ |/* empty */{ $$ = nullptr; };
+
+ for_locking_clause_waiting_policy: SKIP LOCKED{
+   $$ = 0;
+ }
+ | NOWAIT{
+   $$ = 1;
+ }
+ |/* empty */{ $$ = 2; };
```
Unfortunately, integers as representatives for certain modes are not particularly descriptive. As mentioned above, I think using enumeration values would increase the readability.

mweisgut commented on src/sql/SelectStatement.h:
```diff
@@ -96,6 +105,7 @@ struct SelectStatement : SQLStatement {
   std::vector<OrderDescription*>* order;
   std::vector<WithDescription*>* withDescriptions;
   LimitDescription* limit;
+  ForLockingClause* lockings;
```
The implementation of this pull request only supports one locking clause. However, multiple can be passed. Thus, we should store multiple locking clauses here.

https://www.postgresql.org/docs/14/sql-select.html

chrisxu333 replied:
Thank you for accepting my patch into the project. I'm working on refining it now. One question here: what do you mean by saying multiple locking clause can be passed here, could you give an SQL query example of this?

chrisxu333 replied:
I think I figured out what you meant. I'm working on it right now.

mweisgut replied:
The following query is an example with multiple locking clauses:
```
CREATE TABLE t1 (c1 INT);
CREATE TABLE t2 (c1 INT);
INSERT INTO t1 (c1) VALUES (5);
INSERT INTO t2 (c1) VALUES (3);
SELECT t1.c1 AS t1c1, t2.c1 AS t2c1 FROM t1, t2
FOR UPDATE OF t1 NOWAIT
FOR NO KEY UPDATE OF t2 NOWAIT;
```

Verified with PostgreSQL on https://sqliteonline.com/.

mweisgut commented on src/parser/bison_parser.y:
```diff
@@ -746,25 +749,28 @@ select_within_set_operation_no_parentheses : select_clause { $$ = $1; }
   }
 }
-| select_clause set_operator select_within_set_operation opt_order opt_limit {
+| select_clause set_operator select_within_set_operation opt_order for_locking_clause opt_limit {
```
Furthermore, multiple locking clauses are possible in a select statement.

mweisgut reviewed (changes requested):
Thank you, @chrisxu333, for this patch. We want to add your patch to the parser. To do so, we kindly ask you to address the comments listed below.

Besides these aspects, tests are entirely missing. Could you please add tests to `select_tests.cpp`, add correct select statements with the added keywords and right syntax to `queries-good.sql`, and statements with the added keywords buy incorrect syntax to `queries-bad.sql`?
Furthermore, could you please run the `fromat.sh` script from the project's root directory to match the code base's formatting?

chrisxu333 commented:
@mweisgut I‘ve addressed the issues you brought up in your comment in the latest commit, including testcases, format issue, naming conventions as well as multiple locking clause support and such. Please take a look and let me know if anything else needs to be changed. Thank you!

mweisgut commented on src/sql/SelectStatement.h:
```diff
@@ -10,6 +10,9 @@ enum OrderType { kOrderAsc, kOrderDesc };
 
 enum SetType { kSetUnion, kSetIntersect, kSetExcept };
 
+enum LockMode { ForUpdate, ForNoKeyUpdate, ForShare, ForKeyShare };
+enum LockWaitPolicy { NoWait, SkipLocked, None };
```
How about `RowLockMode` and `RowLockWaitPolicy`?

mweisgut commented on src/parser/bison_parser.y:
```diff
@@ -1114,6 +1129,44 @@ alias : AS IDENTIFIER { $$ = new Alias($2); }
+  $$ = new LockingClause();
+  $$->depTable = nullptr;
+  $$->lockMode = $2;
+  $$->lockWaitPolicy = $3;
```
`lockWaitPolicy` --> `rowLockWaitPolicy`?

mweisgut commented on src/sql/SelectStatement.h:
```diff
@@ -56,6 +59,12 @@ struct SetOperation {
+struct LockingClause {
+  LockMode lockMode;
+  LockWaitPolicy lockWaitPolicy;
+  std::vector<char*>* depTable;
```
`tables` or `target_tables` instead of `depTable`?

mweisgut commented on src/util/sqlhelper.cpp:
```diff
@@ -193,6 +193,34 @@ void printSelectStatementInfo(const SelectStatement* stmt, uintmax_t numIndent)
+        inprint("FOR KEY SHARE", numIndent + 3);
+      }
+      if (lockingClause->depTable != nullptr) {
+        inprint("Target table:", numIndent + 2);
```
`Target tables` (plural)?

mweisgut commented on test/queries/queries-bad.sql:
```diff
@@ -41,3 +41,7 @@
+!SELECT * FROM test WHERE val = 2 FOR KEY UPDATE;
+!SELECT * FROM test WHERE val = 2 FOR SHARE test1;
+!SELECT * FROM test WHERE val = 2 FOR NO KEY SHARE;
+!SELECT * FROM test WHERE val = 2 NOWAIT FOR UPDATE;
```
Missing newline at end of file

mweisgut reviewed (changes requested):
Haven't reviewed the tests yet.

chrisxu333 commented:
@mweisgut Thank you for your review. I've pushed another commit and addressed them accordingly. Let me know if the testcases needs to change.

mweisgut commented on test/select_tests.cpp:
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
     }
   }
 }
+TEST(LockingClauseWithoutWaitPolicy) {
```
Please add an empty line between the closing brace and the new test case.

mweisgut replied:
Please also check
- the number of elements in `lockings`
- `tables` (of the struct `LockingClause`)

mweisgut commented on test/select_tests.cpp:
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::None);
+}
+
+TEST(LockingClauseWithWaitPolicy) {
```
Please also check
- the number of elements in `lockings`
- `tables` (of the struct `LockingClause`)

mweisgut commented on test/select_tests.cpp:
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
+  stmt = (SelectStatement*)result.getStatement(7);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForKeyShare);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::SkipLocked);
+}
```
Add an empty line between test cases.

mweisgut commented on test/select_tests.cpp:
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockMode, RowLockMode::ForKeyShare);
+  ASSERT_EQ(stmt->lockings->at(0)->rowLockWaitPolicy, RowLockWaitPolicy::SkipLocked);
+}
+TEST(LockingClauseWithTableReference) {
```
Please also check
- the number of elements in `lockings`
- the number of elements in `tables`

mweisgut commented on test/select_tests.cpp:
```diff
@@ -805,3 +805,123 @@ TEST(IntervalLiteral) {
+  ASSERT_STREQ(stmt->lockings->at(0)->tables->at(1), "c");
+}
+
+TEST(MultipleLockingClause) {
```
Please also check the number of elements in `lockings` and add a statement with multiple locking clauses where one clause has no tables specified, and two other clauses have tables specified but the tables and numbers of specified tables are different.

mweisgut reviewed (changes requested):
A few minor change requests.

chrisxu333 commented:
I've added another commit to cover the change request about testcases. Plz take a look.

mweisgut commented:
@chrisxu333 thank you again for your contribution.
